tools/integration-test/src/tests/client_expiration.rs

```diff
@@ -0,0 +1,458 @@
+                &connection_id_b.as_ref(),
+                &port_a.as_ref(),
+                &port_b.as_ref(),
+            )?;
+
+            // At this point the misbehavior task may raise error, because it
+            // try to check on a client update event that is already expired.
+            // This happens because the misbehavior task is only started when
+            // there is at least one channel in it, _not_ when the client
+            // is created.
+            //
+            // Source of error:
+            // https://github.com/informalsystems/tendermint-rs/blob/c45ea8c82773de1946f7ae2eece13150f07ca5fe/light-client/src/light_client.rs#L216-L222
+
+            assert_eventually_channel_established(
+                &chains.handle_b,
+                &chains.handle_a,
+                &channel_id_b_2.as_ref(),
+                &port_b.as_ref(),
+            )?;
+        }
+
+        Ok(())
+    }
+}
+
+impl BinaryChainTest for PacketExpirationTest {
+    fn run<ChainA: ChainHandle, ChainB: ChainHandle>(
+        &self,
+        _config: &TestConfig,
+        chains: ConnectedChains<ChainA, ChainB>,
+    ) -> Result<(), Error> {
+        let channels = {
+            let _refresh_task_a = spawn_refresh_client(chains.client_b_to_a.clone())
+                .ok_or_else(|| eyre!("expect refresh task spawned"))?;
+
+            let _refresh_task_b = spawn_refresh_client(chains.client_a_to_b.clone())
+                .ok_or_else(|| eyre!("expect refresh task spawned"))?;
+
+            bootstrap_channel_with_chains(&chains, &PortId::transfer(), &PortId::transfer(), false)?
+        };
+
+        wait_for_client_expiry();
+
+        let _supervisor =
+            spawn_supervisor(chains.config.clone(), chains.registry.clone(), None, false)?;
+
+        let denom_a = chains.node_a.denom();
+        let balance_a = chains
+            .node_a
+            .chain_driver()
+            .query_balance(&chains.node_a.wallets().user1().address(), &denom_a)?;
+
+        let denom_b = derive_ibc_denom(
+            &channels.port_b.as_ref(),
+            &channels.channel_id_b.as_ref(),
+            &denom_a,
+        )?;
+
+        {
+            info!("sending first IBC transfer after client is expired. this should cause packet worker to fail");
+
+            chains.node_a.chain_driver().transfer_token(
+                &channels.port_a.as_ref(),
+                &channels.channel_id_a.as_ref(),
+                &chains.node_a.wallets().user1().address(),
+                &chains.node_b.wallets().user1().address(),
+                100,
+                &chains.node_a.denom(),
+            )?;
+
+            sleep(Duration::from_secs(10));
+
+            let balance_a_2 = chains
+                .node_a
+                .chain_driver()
+                .query_balance(&chains.node_a.wallets().user1().address(), &denom_a)?;
+
+            assert_eq(
+                "balance on wallet A should decrease",
```

Comment:
Is that really the behavior we expect? Should we not check whether or not the client is expired before we proceed to the transfer? What happens to the funds withdrawn from wallet A otherwise?

Reply:
The test passes, so that is the current behavior of ibc-go. I think we could file an issue in ibc-go to fail an ibc transfer if the local client is expired or frozen.

On the other hand, a chain probably can't check or guarantee that a counter party client is unfrozen, since everything is asynchronous. So the check would still only handle the simple case.

Reply:
I see! Yeah I think it might be worth filing an issue to discuss this further :)

Reply:
Hmm wait actually it might be that big of a deal, since the IBC packets have timeout. So any such failed token transfer will eventually be refunded. We should also write integration tests for such cases in the future.

Reply:
> The test passes, so that is the current behavior of ibc-go. I think we could file an issue in ibc-go to fail an ibc transfer if the local client is expired or frozen.

I think ibc-go v1.0.0 has protections against this. What chain binary are we using, does it have ibc-go v1.0?